cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MetadataDataset.java

```diff
@@ -152,56 +148,57 @@ private void setMetadata(MetadataEntry metadataEntry) {
 
   @VisibleForTesting
   void setMetadata(MetadataEntry metadataEntry, Set<Indexer> indexers) {
-    NamespacedEntityId targetId = metadataEntry.getTargetId();
-    write(targetId, metadataEntry, indexers);
+    MetadataEntity metadataEntity = metadataEntry.getMetadataEntity();
+    write(metadataEntity, metadataEntry, indexers);
   }
 
   /**
-   * Sets a metadata property for the specified {@link NamespacedEntityId}.
+   * Sets a metadata property for the specified {@link MetadataEntity}.
    *
-   * @param targetId The target Id: {@link ApplicationId} / {@link ProgramId} / {@link DatasetId}/ {@link StreamId}
+   * @param metadataEntity The target Id: {@link ApplicationId} / {@link ProgramId} /
+   * {@link DatasetId}/ {@link StreamId}
    * @param key The metadata key to be added
    * @param value The metadata value to be added
    */
-  public void setProperty(NamespacedEntityId targetId, String key, String value) {
-    setMetadata(new MetadataEntry(targetId, key, value));
+  public void setProperty(MetadataEntity metadataEntity, String key, String value) {
+    setMetadata(new MetadataEntry(metadataEntity, key, value));
   }
 
   /**
-   * Replaces existing tags of the specified {@link NamespacedEntityId} with a new set of tags.
+   * Replaces existing tags of the specified {@link MetadataEntity} with a new set of tags.
    *
-   * @param targetId The target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
+   * @param metadataEntity The target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
    *                 dataset-id(ns+dataset)/stream-id(ns+stream)
    * @param tags the tags to set
    */
-  private void setTags(NamespacedEntityId targetId, String ... tags) {
-    setMetadata(new MetadataEntry(targetId, TAGS_KEY, Joiner.on(TAGS_SEPARATOR).join(tags)));
+  private void setTags(MetadataEntity metadataEntity, String ... tags) {
+    setMetadata(new MetadataEntry(metadataEntity, TAGS_KEY, Joiner.on(TAGS_SEPARATOR).join(tags)));
   }
 
   /**
-   * Adds a new tag for the specified {@link NamespacedEntityId}.
+   * Adds a new tag for the specified {@link MetadataEntity}.
    *
-   * @param targetId the target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
+   * @param metadataEntity the target Id: app-id(ns+app) / program-id(ns+app+pgtype+pgm) /
    *                 dataset-id(ns+dataset)/stream-id(ns+stream).
    * @param tagsToAdd the tags to add
    */
-  public void addTags(NamespacedEntityId targetId, String ... tagsToAdd) {
-    Set<String> existingTags = getTags(targetId);
+  public void addTags(MetadataEntity metadataEntity, String ... tagsToAdd) {
+    Set<String> existingTags = getTags(metadataEntity);
     Iterable<String> newTags = Iterables.concat(existingTags, Arrays.asList(tagsToAdd));
```

Comment:
is this correct? Don't we have to deduplicate? For example, if existing tags is "hello you" and the tags to add are "well hello", then the new list should be "hello you well" and not "hello you well hello"

Reply:
I think this is independent of whether the tag values can contain blanks. It does not help that existingTags is a set, because you convert that to an iterable before appending the new tags. If the existing tags are "hello" and "you", and the new tags are "well" and "hello", then 

- the existingTags set is {"hello", "you"}. 
- Arrays.asList(tagsToAdd) is ["well", "hello"]
- newTags = Iterables.concat(existingTags, that) is ["hello", "you", "well", "hello"]
- join(newTags) is "hello you well hello"

If you don't believe it, please write a test case. There should anyway be a test case that tests deduplication of tags. If there is not then please add one.